Fix static initialisation / destruction issue for Dubins path types

srmainwaring opened the pull request:
Provide safe initialisation and no-destruct for `dubinsPath` object with static storage.

- Closes: https://github.com/ompl/ompl/issues/1270

**Edit** Convert to draft as getting a seg fault on release build on Ubuntu 22.04 in docker. No issue on macOS.

mamoll commented on src/ompl/base/spaces/src/DubinsStateSpace.cpp:
```diff
@@ -712,21 +712,26 @@ namespace ompl::base
-    {{DUBINS_LEFT, DUBINS_RIGHT, DUBINS_LEFT}}
-  }};
+const std::vector<std::vector<DubinsStateSpace::DubinsPathSegmentType> >& DubinsStateSpace::dubinsPathType() {
+  static std::vector<std::vector<DubinsStateSpace::DubinsPathSegmentType> >* pathType
```
Can't this be `std::vector<std::vector<DubinsStateSpace::DubinsPathSegmentType> >` (to avoid heap allocation) and return `pathType`?

srmainwaring replied:
The idea discussed in this article (https://ppwwyyxx.com/blog/2023/Safe-Static-Initialization-No-Destruction), is that the destructor is never called, to ensure that the static storage object does not go out of scope for any other objects referencing it at library unload. Not using a pointer will result in the same issue we currently have, unless we use the [`NoDestructor<>`](https://chromium.googlesource.com/chromium/src/base/+/40343e3fbc1e4835145f125adedcf15a2eb4542f/no_destructor.h) wrapper referred to, which we could do at the cost of introducing the extra template class. I though the pointer to be the simpler of the two, but if heap is a concern I can add the template.

srmainwaring replied:
This is what the `NoDestructor<>` version would look like:
https://github.com/srmainwaring/ompl/tree/prs/pr-dubins-path-static-init-v2

The `NoDestructor<>` is reproduced directly from the link above, I've included the original copyright notice - may need updating to comply with the copyright / license notice for OMPL.

mamoll replied:
ok, I only skimmed the article you linked to the first time. I read it more carefully now. This solution makes sense. There is not a big additional benefit to the `NoDestructor<>` solution, so I'm fine with merging this.

mamoll reviewed (approved):
I'll merge if you remove the "draft" status.

mamoll commented:
One possible change to consider: change this signature:
```
DubinsPath(const std::vector<DubinsPathSegmentType> *type = &dubinsPathType()[0],
                  double t = 0., double p = std::numeric_limits<double>::max(), double q = 0.);
```
to this:
```
DubinsPath(const std::vector<DubinsPathSegmentType>& type = dubinsPathType()[0],
                  double t = 0., double p = std::numeric_limits<double>::max(), double q = 0.);
```
(I.e., use reference rather than a raw C-style pointer.)

srmainwaring commented:
> One possible change to consider: change this signature:

Good suggestion. Updated in https://github.com/ompl/ompl/pull/1271/commits/be319bc6250716764d3c2b1826244d4e257a4cdf.

> I'll merge if you remove the "draft" status.

I moved to draft as I am seeing a segmentation fault in my Ubuntu 22.04 docker build. This may be because I am using `clang++` to build on `x86_64`, so I'll test with `gcc`. 

In the meanwhile, would it be possible to have a very simple `pytest` included in the CI to catch the type of error this PR is trying to resolve. I'm checking locally on macOS, but am concerned that my docker build is not exactly reproducing the CI configuration.

srmainwaring commented:
> This is concerning, because I had to use clang++ on Linux

I've built and run the Dockerfile `./scripts/docker/ompl.Dockerfile`, and the resulting bindings are not exhibiting any issue, so not quite sure what is going on with my image. But this change seems ok.

I had to do the following to get the image to build and work (using Docker Desktop on macOS)

```bash
# build the image
cd  ~/ompl
cp ./scripts/docker/ompl.Dockerfile ./Dockerfile
docker build -t ompl:latest . 

# run the image as root to be able to install python3-pytest
docker run -it --rm -u 0 ompl:latest 
cat > test_import.py <<EOL
def test_import():
    from ompl import base as ob
EOL
apt install python3-pytest
pytest ./test_import.py
```
